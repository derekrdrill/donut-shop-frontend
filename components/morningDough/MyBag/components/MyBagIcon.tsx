import * as React from 'react';
import { useRouter } from 'next/router';
import { Grid } from '@mui/material';

import GlobalContext from '../../../../context/GlobalContext';

import { setAlertItem } from '../../../../components/common/Alert/actions/AlertActions';

import { MyBagIconSVG, MyBagIconBadge, MyBagIconContainer } from '../style';

import { MyBagItem } from '../../../../context/types/MyBagItem';

type MyBagIconProps = {
  color: string;
};

export const getMyBagQuantity = (myBag: Array<MyBagItem>) =>
  myBag.map(myBagItem => myBagItem.quantity).reduce((partialSum, a) => partialSum + a, 0);

const MyBagIcon = ({ color }: MyBagIconProps) => {
  const router = useRouter();

  const {
    dispatch,
    state: { myBag },
  } = React.useContext(GlobalContext);

  return (
    <Grid
      container
      onClick={
        /* istanbul ignore next */
        async () => {
          myBag.length === 0
            ? dispatch(await setAlertItem('No items in your bag', true, 'warning'))
            : router.push('/myBag');
        }
      }
    >
      <MyBagIconContainer item xs={12}>
        <MyBagIconBadge badgeContent={getMyBagQuantity(myBag)} color='secondary' />
        <MyBagIconSVG sx={{ color: color }} />;
      </MyBagIconContainer>
    </Grid>
  );
};

export default MyBagIcon;
