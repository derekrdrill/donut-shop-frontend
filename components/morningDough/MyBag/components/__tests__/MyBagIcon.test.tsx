jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
  useParams: () => ({
    menuItemID: 'chocolateDonut',
  }),
}));

import '@testing-library/jest-dom';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import MyBagIcon, { getMyBagQuantity } from '../MyBagIcon';

describe('MyBag unit tests', () => {
  const myBag = [
    {
      bottled: false,
      butter: null,
      category: 'donuts',
      creamCheese: null,
      dairy: null,
      flavor: null,
      ice: 'normal',
      menuItemID: 'chocolateDonut',
      menuItemImage: 'http://localhost:3000/chocolateDonut.png',
      name: 'Chocolate',
      orderID: '1681765864048-menuItemID',
      quantity: 5,
      size: 'md',
      soda: false,
      subCategory: 'donuts',
      sugar: false,
    },
    {
      bottled: false,
      butter: null,
      category: 'drink',
      creamCheese: null,
      dairy: 'cream',
      flavor: 'frenchVanilla',
      ice: 'normal',
      menuItemID: 'hotCoffee',
      menuItemImage: 'http://localhost:3000/hotCoffee.png',
      name: 'Hot Coffee',
      orderID: '1681765872483-menuItemID',
      quantity: 3,
      size: 'md',
      soda: false,
      subCategory: 'hotDrinks',
      sugar: true,
    },
  ];

  it('renders as expected', () => {
    const result = renderer.create(<MyBagIcon color='gainsboro' />).toJSON();
    expect(result).toMatchSnapshot();
  });

  it('runs getMyBagQuantity correctly', () => {
    expect(getMyBagQuantity(myBag)).toEqual(8);
  });
});
