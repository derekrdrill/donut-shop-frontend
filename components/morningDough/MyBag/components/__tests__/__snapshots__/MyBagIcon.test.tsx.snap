// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyBag unit tests renders as expected 1`] = `
<div
  className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
  onClick={[Function]}
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 sc-iAEyYk bDdAOG css-1idn90j-MuiGrid-root"
  >
    <span
      className="MuiBadge-root sc-dmqHEX djqssH css-1c32n2y-MuiBadge-root"
    >
      <span
        className="MuiBadge-badge MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorSecondary css-fu0map-MuiBadge-badge"
      />
    </span>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-gueYoa eACLnc css-t1n1uy-MuiSvgIcon-root"
      data-testid="ShoppingBagIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M18 6h-2c0-2.21-1.79-4-4-4S8 3.79 8 6H6c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2zm-8 4c0 .55-.45 1-1 1s-1-.45-1-1V8h2v2zm2-6c1.1 0 2 .9 2 2h-4c0-1.1.9-2 2-2zm4 6c0 .55-.45 1-1 1s-1-.45-1-1V8h2v2z"
      />
    </svg>
    ;
  </div>
</div>
`;
